File: services/workflow-orchestrator/activities/log_external_event.py
"""
Log External Event Activity

Persists external event records (approval requests, responses, timeouts)
to the database for audit trail purposes.

Uses Dapr service invocation to call function-router which has database access.
"""

from __future__ import annotations

import logging
import os
from typing import Any, Literal

import requests

logger = logging.getLogger(__name__)

DAPR_HOST = os.environ.get("DAPR_HOST", "localhost")
DAPR_HTTP_PORT = os.environ.get("DAPR_HTTP_PORT", "3500")
FUNCTION_ROUTER_APP_ID = os.environ.get("FUNCTION_ROUTER_APP_ID", "function-router")

ExternalEventType = Literal["approval_request", "approval_response", "timeout"]


def log_external_event(ctx, input_data: dict[str, Any]) -> dict[str, Any]:
    """
    Log an external event to the database.

    This activity calls the function-router's /external-event endpoint
    to persist the event record.

    Args:
        ctx: Dapr workflow context (not used but required by Dapr)
        input_data: Dict with event details

    Returns:
        Dict with success status and event_id
    """
    execution_id = input_data.get("executionId", "")
    node_id = input_data.get("nodeId", "")
    event_name = input_data.get("eventName", "")
    event_type = input_data.get("eventType", "")

    logger.info(
        f"[Log External Event] Logging {event_type} for event: {event_name} "
        f"(execution: {execution_id})"
    )

    try:
        url = f"http://{DAPR_HOST}:{DAPR_HTTP_PORT}/v1.0/invoke/{FUNCTION_ROUTER_APP_ID}/method/external-event"

        request_payload = {
            "execution_id": execution_id,
            "node_id": node_id,
            "event_name": event_name,
            "event_type": event_type,
            "timeout_seconds": input_data.get("timeoutSeconds"),
            "approved": input_data.get("approved"),
            "reason": input_data.get("reason"),
            "responded_by": input_data.get("respondedBy"),
            "payload": input_data.get("payload"),
        }

        response = requests.post(url, json=request_payload, timeout=30)
        response.raise_for_status()

        result = response.json()

        if not result.get("success"):
            logger.warning(
                f"[Log External Event] Failed to log {event_type}: {result.get('error')}"
            )
            return {
                "success": False,
                "error": result.get("error"),
            }

        logger.info(
            f"[Log External Event] Successfully logged {event_type}: {result.get('event_id')}"
        )

        return {
            "success": True,
            "eventId": result.get("event_id"),
        }

    except Exception as e:
        error_msg = str(e)
        logger.error(f"[Log External Event] Error logging {event_type}: {e}")

        # Don't throw - audit logging failure shouldn't break workflow execution
        return {
            "success": False,
            "error": f"Failed to log external event: {error_msg}",
        }


def log_approval_request(ctx, input_data: dict[str, Any]) -> dict[str, Any]:
    """Helper to log an approval request event."""
    return log_external_event(ctx, {
        "executionId": input_data.get("executionId"),
        "nodeId": input_data.get("nodeId"),
        "eventName": input_data.get("eventName"),
        "eventType": "approval_request",
        "timeoutSeconds": input_data.get("timeoutSeconds"),
    })


def log_approval_response(ctx, input_data: dict[str, Any]) -> dict[str, Any]:
    """Helper to log an approval response event."""
    return log_external_event(ctx, {
        "executionId": input_data.get("executionId"),
        "nodeId": input_data.get("nodeId"),
        "eventName": input_data.get("eventName"),
        "eventType": "approval_response",
        "approved": input_data.get("approved"),
        "reason": input_data.get("reason"),
        "respondedBy": input_data.get("respondedBy"),
        "payload": input_data.get("payload"),
    })


def log_approval_timeout(ctx, input_data: dict[str, Any]) -> dict[str, Any]:
    """Helper to log a timeout event."""
    timeout_seconds = input_data.get("timeoutSeconds", 0)
    return log_external_event(ctx, {
        "executionId": input_data.get("executionId"),
        "nodeId": input_data.get("nodeId"),
        "eventName": input_data.get("eventName"),
        "eventType": "timeout",
        "timeoutSeconds": timeout_seconds,
        "reason": f"Timed out after {timeout_seconds} seconds",
    })

File: services/workflow-orchestrator-ts-archived/src/activities/log-external-event.ts
/**
 * Log External Event Activity
 *
 * Persists external event records (approval requests, responses, timeouts)
 * to the database for audit trail purposes.
 *
 * Uses Dapr service invocation to call function-router which has database access.
 */
import { DaprClient, HttpMethod } from "@dapr/dapr";

const DAPR_HOST = process.env.DAPR_HOST || "localhost";
const DAPR_HTTP_PORT = process.env.DAPR_HTTP_PORT || "3500";
const FUNCTION_ROUTER_APP_ID =
  process.env.FUNCTION_ROUTER_APP_ID || "function-router";

/**
 * External event types
 */
export type ExternalEventType =
  | "approval_request"
  | "approval_response"
  | "timeout";

/**
 * Input for logging an external event
 */
export type LogExternalEventInput = {
  executionId: string; // Database execution ID
  nodeId: string;
  eventName: string;
  eventType: ExternalEventType;
  timeoutSeconds?: number;
  approved?: boolean;
  reason?: string;
  respondedBy?: string;
  payload?: Record<string, unknown>;
};

/**
 * Output from logging an external event
 */
export type LogExternalEventOutput = {
  success: boolean;
  eventId?: string;
  error?: string;
};

/**
 * Log an external event to the database
 *
 * This activity calls the function-router's /external-event endpoint
 * to persist the event record.
 */
export async function logExternalEvent(
  _ctx: unknown,
  input: LogExternalEventInput
): Promise<LogExternalEventOutput> {
  const { executionId, nodeId, eventName, eventType } = input;

  console.log(
    `[Log External Event] Logging ${eventType} for event: ${eventName} (execution: ${executionId})`
  );

  try {
    const client = new DaprClient({
      daprHost: DAPR_HOST,
      daprPort: DAPR_HTTP_PORT,
    });

    const requestPayload = {
      execution_id: executionId,
      node_id: nodeId,
      event_name: eventName,
      event_type: eventType,
      timeout_seconds: input.timeoutSeconds,
      approved: input.approved,
      reason: input.reason,
      responded_by: input.respondedBy,
      payload: input.payload,
    };

    const result = (await client.invoker.invoke(
      FUNCTION_ROUTER_APP_ID,
      "external-event",
      HttpMethod.POST,
      requestPayload
    )) as { success: boolean; event_id?: string; error?: string };

    if (!result.success) {
      console.warn(
        `[Log External Event] Failed to log ${eventType}: ${result.error}`
      );
      return {
        success: false,
        error: result.error,
      };
    }

    console.log(
      `[Log External Event] Successfully logged ${eventType}: ${result.event_id}`
    );

    return {
      success: true,
      eventId: result.event_id,
    };
  } catch (error) {
    const errorMessage = error instanceof Error ? error.message : String(error);

    console.error(`[Log External Event] Error logging ${eventType}:`, error);

    // Don't throw - audit logging failure shouldn't break workflow execution
    return {
      success: false,
      error: `Failed to log external event: ${errorMessage}`,
    };
  }
}

/**
 * Helper to log an approval request event
 */
export async function logApprovalRequest(
  _ctx: unknown,
  input: {
    executionId: string;
    nodeId: string;
    eventName: string;
    timeoutSeconds: number;
  }
): Promise<LogExternalEventOutput> {
  return logExternalEvent(null, {
    executionId: input.executionId,
    nodeId: input.nodeId,
    eventName: input.eventName,
    eventType: "approval_request",
    timeoutSeconds: input.timeoutSeconds,
  });
}

/**
 * Helper to log an approval response event
 */
export async function logApprovalResponse(
  _ctx: unknown,
  input: {
    executionId: string;
    nodeId: string;
    eventName: string;
    approved: boolean;
    reason?: string;
    respondedBy?: string;
    payload?: Record<string, unknown>;
  }
): Promise<LogExternalEventOutput> {
  return logExternalEvent(null, {
    executionId: input.executionId,
    nodeId: input.nodeId,
    eventName: input.eventName,
    eventType: "approval_response",
    approved: input.approved,
    reason: input.reason,
    respondedBy: input.respondedBy,
    payload: input.payload,
  });
}

/**
 * Helper to log a timeout event
 */
export async function logApprovalTimeout(
  _ctx: unknown,
  input: {
    executionId: string;
    nodeId: string;
    eventName: string;
    timeoutSeconds: number;
  }
): Promise<LogExternalEventOutput> {
  return logExternalEvent(null, {
    executionId: input.executionId,
    nodeId: input.nodeId,
    eventName: input.eventName,
    eventType: "timeout",
    timeoutSeconds: input.timeoutSeconds,
    reason: `Timed out after ${input.timeoutSeconds} seconds`,
  });
}
